### PR TITLE
add conviction policy interface

### PR DESCRIPTION
**Pull request by changliuau**
#1079 

Conviction policy: 

decide when to mark the host as 'down' based on the error. Default:  `SimpleConvictionPolicy` - mark host as 'down' on any error.

Opened interface in policies.go for user to implement their own policies.

### CODE REVIEW

File: policies.go [hunk 1]
```diff
@@ -706,3 +706,20 @@ func (d *dcAwareRR) Pick(q ExecutableQuery) NextHost {
 		return (*selectedHost)(host)
 	}
 }
+
+type ConvictionPolicy interface {
+	// Implementations should return `true` if the host should be convicted, `false` otherwise.
+	AddFailure(error error) bool
```
Comment (Zariel) [inline policies.go]: Surely this should take the host info that caused the error?

File: policies.go [hunk 2]
```diff
@@ -791,6 +791,22 @@ func (d *dcAwareRR) Pick(q ExecutableQuery) NextHost {
 	}
 }
 
+type ConvictionPolicy interface {
+	// Implementations should return `true` if the host should be convicted, `false` otherwise.
+	AddFailure(error error, host *HostInfo) bool
+	//Implementations should clear out any convictions or state regarding the host.
+	Reset()
```
Comment (Zariel) [inline policies.go]: Should this not also take a host given the comment?

File: policies.go [hunk 3]
```diff
@@ -791,6 +791,22 @@ func (d *dcAwareRR) Pick(q ExecutableQuery) NextHost {
...
+}
+
+//return true on any error
```
Comment (Zariel) [inline policies.go]: Improve this comment, something like
`SimpleConvictionPolicy implements a ConvictionPolicy which convicts all hosts regardless of error`

File: policies.go [hunk 4]
```diff
@@ -791,6 +791,22 @@ func (d *dcAwareRR) Pick(q ExecutableQuery) NextHost {
 	}
 }
 
+type ConvictionPolicy interface {
```
Comment (Zariel) [inline policies.go]: Also add a comment explaining what this interface is for.

**Comment by changliuau** [comment]
@Zariel  thanks - fixed the comments and Reset function as suggested

**Pull request merged**